runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -1859,15 +2177,42 @@ J9::TransformUtil::transformIndirectLoadChainImpl(TR::Compilation *comp, TR::Nod
       return false;
       }
 
-   // Dereference the chain starting from baseAddress and get the field address
-   void *fieldAddress = dereferenceStructPointerChain(baseAddress, baseExpression, isBaseStableArray, node, comp);
-   if (!fieldAddress)
+   // in non-jitserver mode, we need to hold a mutex until the end of the function to ensure
+   // fieldAddress holds
+   TR::VMAccessCriticalSection transformIndirectLoadChain(comp,
```

Comment:
Is there some other way of writing a critical section that will work for the entire function scope, but only for non-jitserver mode?

Reply:
The server always has VM access in hand because GC is not active and there is no risk of slowing it down:
```
      if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
         {
         // The server can compile with VM access in hand because GC is not a factor here
         // For the same reason we don't have to use TR_EnableYieldVMAccess
         self()->setOption(TR_DisableNoVMAccess);
```
This means that you can use the same VMaccess critical section for both non-jitserver and jitserver. For jitserver it's going to be a no-op since it already has VM access